File: xeng_sim/make_spead_data.py
__author__ = 'paulp'

valid_ant = 0
valid_pol = 0
start_time = 83545080 << (9 + 1 + 7)
start_time = 1 << (9 + 1 + 7)
time_increment = 1 << (9 + 1 + 7)
num_timestamps_to_process = 30

# constants
start_from = 113
valid_gap = 65  # if this is too small if won't work! - the reorder block falls over
# correlator setup
num_ant = 4
num_x = 32
num_chans = 512
xacc_len = 128
# calculated
chans_per_x = num_chans / num_x
spead_packet_len = xacc_len / 2
window_len = spead_packet_len * num_ant * chans_per_x

# print a little summary:
print 'Channels: ', num_chans
print 'f-engines: ', num_ant
print 'x-engines: ', num_x
print 'Chans per x: ', chans_per_x
print 'X-acc length: ', xacc_len
print 'Gap between valid blocks: ', valid_gap
print 'SPEAD packet length: ', spead_packet_len
print 'Window length: ', window_len
print ''


def make_spead_data(packet_length, data_value, packet_time, input_id, channel):
    spead_data = {'data':       [data_value] * packet_length,
                  'valid':      [1] * packet_length,
                  'eof':        [0] * packet_length,
                  'time48':     [packet_time] * packet_length,
                  'fengid':     [input_id] * packet_length,
                  'flags':      [0] * packet_length,
                  'channel':    [channel] * packet_length,
                  }
    spead_data['eof'][-1] = 1
    return spead_data


# number of timestamps to do - make all the spead packets
packets = []
block_position = 0
freq_block_size = window_len / chans_per_x
for timectr in range(0, num_timestamps_to_process):
    this_time = start_time + (timectr * time_increment)
    # print 'time %i starts at %i:' % (this_time, block_position)
    for freqchan in range(0, num_chans / num_x):
        # freq_position = block_position + (freqchan * freq_block_size)
        # print '\tfreq %i at %i:' % (freqchan, freq_position)
        # print '\t\t',
        for ant in range(0, num_ant):
            # ant_position = freq_position + (ant * spead_packet_len)
            # print 'ant%02i: %i,' % (ant, ant_position),
            # data
            if ant == valid_ant:
                if valid_pol == 0:
                    dataval = 72057594054705152
                else:
                    dataval = 1099511628032
            else:
                dataval = 0
            # make the packet
            pkt = make_spead_data(spead_packet_len, dataval, this_time, ant, freqchan)
            packets.append(pkt)
        # print ''
    block_position += window_len

# make the vectors
packetdata = {}
for key in packets[0].keys():
    packetdata[key] = [0] * start_from

# add the data
for pkt in packets:
    for key in packets[0].keys():
        packetdata[key].extend(pkt[key])
        packetdata[key].extend([0] * valid_gap)

# reorder to columns for matlab and make the matlab dict
import numpy as np
total_length = len(packetdata['valid'])
print 'Total vector length: ', total_length
mat_dict = {}
for key in packets[0].keys():
    mat_dict['sim_spead_%s' % key] = {'time': [],
                                'signals': {'values': np.array(packetdata[key],
                                                               dtype=np.double).reshape(total_length, 1),
                                            'dimensions': 1.0}
    }

# save the matlab dict to a file
import scipy.io as sio
sio.savemat('/tmp/xreord.mat', mat_dict)

# end

File: xeng_sim/make_speaddata.m
sim_len = 50000;
valid_ant = 0;
valid_pol = 0;
start_time = 83545080;
time_increment = 1;

% constants
start_from = 113;
valid_gap = 111;
% correlator setup
num_ant = 4;
num_x = 32;
num_chans = 512;
xacc_len = 128;
% calculated
chans_per_x = num_chans / num_x;
spead_packet_len = xacc_len / 4;

% print a little summary:
fprintf('Channels: %i\n', num_chans);
fprintf('f-engines: %i\n', num_ant);
fprintf('x-engines: %i\n', num_x);
fprintf('Chans per x: %i\n', chans_per_x);
fprintf('X-acc length: %i\n', xacc_len);
fprintf('Gap between valid blocks: %i\n', valid_gap);
fprintf('SPEAD packet length: %i\n', spead_packet_len);
fprintf('\n');

% number of timestamps to do
sizectr = 0;
for timectr = start_time : start_time + 3,
    for freqchan = 1 : num_chans / num_x,
        for ant = 1 : num_ant,
            % headers
            spead_time48() = timectr;
            spead_fengid() = ant;
            spead_flags() = 0;
            spead_freqid() = freqchan;
            % data array
            timeramp = 1 : xacc_len;
            timeramp = reshape(timeramp, 4, xacc_len/4);
            
            
        end
    end
    sizectr = sizectr + 1;
end







% make the main valid window
% valid_block = cat(2, ones(1, block_len), zeros(1, valid_gap));
% valid_block = repmat(valid_block, 1, blocks_per_window);
valid_vector = ones(1, block_len * blocks_per_window);
sync_vector = zeros(1, length(valid_vector));
sync_vector(sync_offset * xacc_len * num_ant) = 1;
datasel_vector = cat(2, zeros(1, xacc_len * valid_ant), ones(1, xacc_len), zeros(1, xacc_len * (num_ant - valid_ant - 1)));
datasel_vector = repmat(datasel_vector, 1, n_consecutive_chans);
datasel_vector = repmat(datasel_vector, 1, blocks_per_window);

valid_vector = reshape(valid_vector, block_len, blocks_per_window);
sync_vector = reshape(sync_vector, block_len, blocks_per_window);
datasel_vector = reshape(datasel_vector, block_len, blocks_per_window);

valid_block2 = zeros(block_len + valid_gap, blocks_per_window);
sync_vector2 = zeros(block_len + valid_gap, blocks_per_window);
datasel_vector2 = zeros(block_len + valid_gap, blocks_per_window);
for ctr = 1 : blocks_per_window,
    valid_block2(:, ctr) = cat(1, valid_vector(:, ctr), zeros(valid_gap, 1));
    sync_vector2(:, ctr) = cat(1, sync_vector(:, ctr), zeros(valid_gap, 1));
    datasel_vector2(:, ctr) = cat(1, datasel_vector(:, ctr), zeros(valid_gap, 1));
end
valid_vector = reshape(valid_block2, 1, blocks_per_window * (block_len + valid_gap));
sync_vector = reshape(sync_vector2, 1, blocks_per_window * (block_len + valid_gap));
datasel_vector = reshape(datasel_vector2, 1, blocks_per_window * (block_len + valid_gap));
clear valid_block2 sync_block2 datasel_vector2;

% repeat them all to the sim length
num_reps = ceil(sim_len / length(valid_vector));
valid_vector = repmat(valid_vector, 1, num_reps);
datasel_vector = repmat(datasel_vector, 1, num_reps);
sync_vector = repmat(sync_vector, 1, num_reps);
clear num_reps;

% make the time vector
time_vector = zeros(1, length(valid_vector));
this_time = start_time;
spos = 1;
sync_locations = find(sync_vector);
for ctr = 1 : length(sync_locations),
    epos = sync_locations(ctr);
    fprintf('%7i:%7i - time %i\n', spos, epos, this_time);
    time_vector(spos : epos) = ones(1, epos - spos + 1) * this_time;
    spos = epos + 1;
    this_time = this_time + time_increment;
end
clear spos this_time;

% make the correct pol select vector
pol_select_vector = datasel_vector * valid_pol;

% f = figure();
% subplot(5,1,1)
% plot(valid_vector, 'r');
% caxes = axis();
% axis([caxes(2)*-0.1 caxes(2)*1.1 caxes(4)*-0.1 caxes(4)*1.1]);
% subplot(5,1,2)
% plot(sync_vector, 'b');
% caxes = axis();
% axis([caxes(2)*-0.1 caxes(2)*1.1 caxes(4)*-0.1 caxes(4)*1.1]);
% subplot(5,1,3)
% plot(datasel_vector, 'g');
% caxes = axis();
% axis([caxes(2)*-0.1 caxes(2)*1.1 caxes(4)*-0.1 caxes(4)*1.1]);
% subplot(5,1,4)
% plot(pol_select_vector, 'c');
% caxes = axis();
% axis([caxes(2)*-0.1 caxes(2)*1.1 caxes(4)*-0.1 caxes(4)*1.1]);
% subplot(5,1,5)
% time_vector_plot = time_vector;
% time_vector_plot(time_vector_plot == 0) = max(time_vector_plot);
% plot(time_vector_plot, 'm');
% caxes = axis();
% axis([caxes(2)*-0.1 caxes(2)*1.1 min(time_vector_plot)-1 max(time_vector_plot)+1]);

% make the simulink structures
sim_valid.time = []; sim_valid.signals.values = valid_vector'; sim_valid.signals.dimensions = 1;
sim_data_sel.time = []; sim_data_sel.signals.values = datasel_vector'; sim_data_sel.signals.dimensions = 1;
sim_pol_sel.time = []; sim_pol_sel.signals.values = pol_select_vector'; sim_pol_sel.signals.dimensions = 1;
sim_sync.time = []; sim_sync.signals.values = sync_vector'; sim_sync.signals.dimensions = 1;
sim_time.time = []; sim_time.signals.values = time_vector'; sim_time.signals.dimensions = 1;
clear valid_vector datasel_vector pol_select_vector sync_vector;

% done
